### PR TITLE
Add mip dna rules

### PR DESCRIPTION
**Pull request by seallard**
Add mip-dna rules (same as for tomte). Fixed type hints using aliases as agreed.

Closes https://github.com/Clinical-Genomics/cg/issues/3614

### CODE REVIEW

File: cg/services/order_validation_service/rules/case/rules.py [hunk 1]
```diff
@@ -44,3 +49,7 @@ def validate_case_internal_ids_exist(
             error = CaseDoesNotExistError(case_index=case_index)
             errors.append(error)
     return errors
+
+
+def validate_case_names_not_repeated(order: TomteOrder, **kwargs) -> list[RepeatedCaseNameError]:
```
[the remark this answers is not in the document]
Comment (seallard) [reply cg/services/order_validation_service/rules/case/rules.py]: https://github.com/Clinical-Genomics/cg/pull/3625/commits/e53fef24b29aa440acd8105b9f11aa0f9d49c005

File: cg/services/order_validation_service/rules/case_sample/utils.py [hunk 2]
```diff
@@ -1,19 +1,48 @@
...
 
 
-def is_concentration_missing(sample: TomteSample) -> bool:
+def is_application_not_compatible(
+    allowed_prep_categories: list[PrepCategory],
+    application_tag: str,
+    store: Store,
+) -> bool:
+    application: Application = store.get_application_by_tag(application_tag)
+    return application and (application.prep_category not in allowed_prep_categories)
```
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:34]: Why is this added here? We have it in order_validation_service/rules/utils.py already

File: cg/services/order_validation_service/rules/case_sample/utils.py [hunk 3]
```diff
@@ -1,19 +1,48 @@
...
+
+
+def is_concentration_missing(sample: SampleWithRelatives) -> bool:
```
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:37]: Probably for the future but this type hint is very confusing. The check has nothing to do with relatives, and there will be pipelines where trios are not submitted but which need this check.

File: cg/services/order_validation_service/rules/case_sample/utils.py [hunk 4]
```diff
@@ -1,19 +1,48 @@
...
 
 
-def is_well_position_missing(sample: TomteSample) -> bool:
+def is_well_position_missing(sample: SampleWithRelatives) -> bool:
     return sample.container == ContainerEnum.plate and not sample.well_position
 
 
-def is_container_name_missing(sample: TomteSample) -> bool:
+def is_container_name_missing(sample: SampleWithRelatives) -> bool:
     return sample.container == ContainerEnum.plate and not sample.container_name
```
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:46]: These should be moved to order_validation_service/rules/utils.py imo

File: cg/services/order_validation_service/rules/case_sample/utils.py [hunk 5]
```diff
@@ -22,3 +51,252 @@ def get_invalid_panels(panels: list[str], store: Store) -> list[str]:
...
+    allowed_volume: bool = is_volume_within_allowed_interval(sample.volume)
+    return in_container and not allowed_volume
+
+
+def is_in_container(container: ContainerEnum) -> bool:
+    return container != ContainerEnum.no_container
+
+
+def is_volume_within_allowed_interval(volume: int) -> bool:
+    return MINIMUM_VOLUME <= volume <= MAXIMUM_VOLUME
```
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:67]: These exist in order_validation_service/rules/utils.py

File: cg/services/order_validation_service/rules/case_sample/utils.py [hunk 6]
```diff
@@ -22,3 +51,252 @@ def get_invalid_panels(panels: list[str], store: Store) -> list[str]:
...
+def get_occupied_well_errors(colliding_samples: list[tuple[int, int]]) -> list[OccupiedWellError]:
+    errors: list[OccupiedWellError] = []
+    for sample_index, case_index in colliding_samples:
+        error = OccupiedWellError(case_index=case_index, sample_index=sample_index)
+        errors.append(error)
+    return errors
+
+
+def is_sample_on_plate(sample: SampleWithRelatives) -> bool:
+    return sample.container == ContainerEnum.plate
```
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:99]: This feels like it should be in order_validation_service/rules/utils.py with a `Sample` type hint

File: cg/services/order_validation_service/rules/case_sample/utils.py [hunk 7]
```diff
@@ -22,3 +51,252 @@ def get_invalid_panels(panels: list[str], store: Store) -> list[str]:
...
+
+    return indices
+
+
+def get_repeated_case_name_errors(order: OrderWithCases) -> list[RepeatedCaseNameError]:
+    case_indices: list[int] = get_indices_for_repeated_case_names(order)
+    return [RepeatedCaseNameError(case_index=case_index) for case_index in case_indices]
+
+
+def get_indices_for_repeated_sample_names(case: CaseContainingRelatives) -> list[int]:
```
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:118]: Could this not make do with a `Case` typehint?

File: cg/services/order_validation_service/rules/case_sample/utils.py [hunk 8]
```diff
@@ -22,3 +51,252 @@ def get_invalid_panels(panels: list[str], store: Store) -> list[str]:
...
+
+    for index, sample in case.enumerated_new_samples:
+        if counter.get(sample.name) > 1:
+            indices.append(index)
+
+    return indices
+
+
+def get_repeated_sample_name_errors(
+    case: CaseContainingRelatives, case_index: int
```
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:130]:
```suggestion
    case: Case, case_index: int
```

File: cg/services/order_validation_service/rules/case_sample/utils.py [hunk 9]
```diff
@@ -22,3 +51,252 @@ def get_invalid_panels(panels: list[str], store: Store) -> list[str]:
...
+
+
+def create_mother_sex_error(case_index: int, sample_index: int) -> InvalidMotherSexError:
+    return InvalidMotherSexError(case_index=case_index, sample_index=sample_index)
+
+
+def validate_subject_ids_in_case(
+    case: CaseContainingRelatives, case_index: int
+) -> list[SubjectIdSameAsCaseNameError]:
+    errors: list[SubjectIdSameAsCaseNameError] = []
```
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:231]: Again, nothing to do with relatives and will probably need another typing but keep it for now.

File: cg/services/order_validation_service/rules/case_sample/utils.py [hunk 10]
```diff
@@ -22,3 +51,252 @@ def get_invalid_panels(panels: list[str], store: Store) -> list[str]:
...
+    for sample_index, sample in case.enumerated_new_samples:
+        if sample.subject_id == case.name:
+            error = SubjectIdSameAsCaseNameError(case_index=case_index, sample_index=sample_index)
+            errors.append(error)
+    return errors
+
+
+def validate_concentration_in_case(
+    case: CaseContainingRelatives, case_index: int, store: Store
```
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:240]: Same thing here. Keep for now, but it will probably be changed

File: cg/services/order_validation_service/workflows/mip_dna/models/sample.py [hunk 11]
```diff
@@ -1,5 +1,24 @@
...
+    age_at_sampling: float | None = None
+    control: ControlEnum | None = None
+    elution_buffer: str | None = None
+    father: str | None = Field(None, pattern=NAME_PATTERN)
+    formalin_fixation_time: int | None = None
+    mother: str | None = Field(None, pattern=NAME_PATTERN)
+    phenotype_groups: list[str] | None = None
+    phenotype_terms: list[str] | None = None
+    post_formalin_fixation_time: int | None = None
+    reference_genome: GenomeVersion
```
Comment (islean) [inline cg/services/order_validation_service/workflows/mip_dna/models/sample.py:18]:
```suggestion
```

File: cg/services/order_validation_service/workflows/mip_dna/validation_rules.py [hunk 12]
```diff
@@ -1,3 +1,47 @@
...
+    validate_sex_required_for_new_samples,
+    validate_source_required,
+    validate_status_required_if_new,
+    validate_subject_ids_different_from_case_names,
+    validate_subject_ids_different_from_sample_names,
+    validate_volume_interval,
+    validate_wells_contain_at_most_one_sample,
+)
+
 CASE_RULES: list[callable] = []
```
Comment (islean) [inline cg/services/order_validation_service/workflows/mip_dna/validation_rules.py:24]: Where are the case rules?

**Review by islean — APPROVED** [review]
Some minor comments! I also think we need a `MipDnaOrder` class to appropriately type hint the delivery type

File: cg/services/order_validation_service/rules/case_sample/utils.py (hunk 2 above)
Comment (seallard) [reply cg/services/order_validation_service/rules/case_sample/utils.py]: https://github.com/Clinical-Genomics/cg/pull/3625/commits/62c1a54290cd0ff16666334e1685b66f4fa34cc7

File: cg/services/order_validation_service/rules/case_sample/utils.py (hunk 4 above)
Comment (seallard) [reply cg/services/order_validation_service/rules/case_sample/utils.py:37]: It operates on a case sample so I'll leave it here for now

File: cg/services/order_validation_service/rules/case_sample/utils.py (hunk 5 above)
Comment (seallard) [reply cg/services/order_validation_service/rules/case_sample/utils.py]: https://github.com/Clinical-Genomics/cg/pull/3625/commits/03d1b7263aab6bb51e2a44ac7d2743115b542ff4

File: cg/services/order_validation_service/rules/case_sample/utils.py (hunk 6 above)
Comment (seallard) [reply cg/services/order_validation_service/rules/case_sample/utils.py]: https://github.com/Clinical-Genomics/cg/pull/3625/commits/6b815e9b024fffa5a9ad039429a0bf2d77bbd36e

File: cg/services/order_validation_service/rules/case_sample/utils.py (hunk 7 above)
Comment (seallard) [reply cg/services/order_validation_service/rules/case_sample/utils.py]: https://github.com/Clinical-Genomics/cg/pull/3625/commits/d4639b908ce95a8fe2881019e3934bd844f94afb

File: cg/services/order_validation_service/workflows/mip_dna/models/sample.py (hunk 11 above)
Comment (seallard) [reply cg/services/order_validation_service/workflows/mip_dna/models/sample.py]: https://github.com/Clinical-Genomics/cg/pull/3625/commits/9c963a89a2cca5703204e562d7f7174563ef27c5

File: cg/services/order_validation_service/workflows/mip_dna/validation_rules.py (hunk 12 above)
Comment (seallard) [reply cg/services/order_validation_service/workflows/mip_dna/validation_rules.py]: https://github.com/Clinical-Genomics/cg/pull/3625/commits/22b77ca773c6e0c4f5ba5235ed4c3ddf20551148
